### DISCUSSION TITLE
Is binaryToJsonString just converting UTF-8 bytes to a string?

### DISCUSSION THREAD

**Discussion by rgl** [anacrolix/torrent#696] [General]
https://github.com/anacrolix/torrent/blob/615ac41ae3e0e88515cf2ff73d6f3ce6e9777420/webtorrent/tracker_protocol.go#L40-L48

@anacrolix, is that function really needed? can't it be implemented simply as `string(b)`?

**Comment by anacrolix** [comment]
No, I don't think so. I've had to do this in a few places in Go. To tunnel raw bytes through JSON strings requires us to force Go to expand characters over 0x7f in to 2 bytes in a string, which JSON then encodes correctly using its \u escaping or something. It's been a while since I've had to meddle with this.

**Reply by rgl** [reply]
By any chance, do you have an example of a source byte array? I'm really curious about this :-)

**Reply by anacrolix** [reply]
Not to hand, try any byte array that contains non-printable characters that JSON strings do not allow without escaping. Above 0x7f as above. I see the function is fed the output of rand.Read, you could try round-tripping that to JSON and back and see if you get the same thing with and without the function. That would make an interesting fuzz target.

**Reply by anacrolix** [reply]
The code was taken from here, which is copied from a JS/Node project: https://github.com/anacrolix/torrent/commit/e5fd9ede549a96e936b1692542b8753dfc1b3f90#diff-60fe39a4b532c5141a978580c841f84d6374928867a74940dc271d3aeeeb7901L42-L48
